File: cpp_prac/kakaothatsong.cpp
#include <string>
#include <vector>

using namespace std;

vector<string> strdeli(string str)
{
    int st=0;
    vector<string> pres;
    for(int i=0; i<str.size(); ++i)
    {
        if(str[i]==',')
        {
            pres.push_back(str.substr(st,i-st));
            st=i+1;
        }
    }
    pres.push_back(str.substr(st));
    return pres;
}
string solution(string m, vector<string> musicinfos) {
    string answer = "";
    int anst=0;
    for(int i=0; i<musicinfos.size(); ++i)
    {
        vector<string> pres=strdeli(musicinfos[i]);
        int fint=stoi(pres[1].substr(0,2))*60+ stoi(pres[1].substr(3,2));
        int stt=stoi(pres[0].substr(0,2))*60+ stoi(pres[0].substr(3,2));
        int intert=fint-stt;
        int slen=pres[3].size();
        string summelo="";
        for(int j=0; j<pres[3].size(); ++j){if(pres[3][j]=='#') slen--;}
        for(int j=0; j<intert/slen; ++j){summelo+=pres[3];}
        int count=intert%slen;
        for(int j=0; count!=0; j++)
        {
            summelo.push_back(pres[3][j]);
            if(pres[3][j]!='#') count--;
        }
        summelo.push_back('0');
        int tt1=summelo.size();
        int tt2=m.size();
        for(int j=0; j<tt1-tt2; ++j)
        {
            if(summelo.substr(j,m.size())==m && summelo[j+m.size()]!='#' && intert>anst)
            {
                answer=pres[2];
                anst=intert;
            }
        }
    }
    if(answer=="") answer="(None)";
    return answer;
}

File: java_prac/kakaothatsong.java
class Solution {
    public String solution(String m, String[] musicinfos) {
        String answer = "";
        int anst=0;
        int ansst=1441;
        for(int i=0; i<musicinfos.length; ++i)
        {
            String[] pres=musicinfos[i].split(",");
            String wholem="";
            int stt=Integer.parseInt(pres[0].substring(0,2))*60 + Integer.parseInt(pres[0].substring(3,5));
            int fit=Integer.parseInt(pres[1].substring(0,2))*60 + Integer.parseInt(pres[1].substring(3,5));
            int intert=fit-stt;
            int slen=pres[3].length();
            for(int j=0; j<pres[3].length(); ++j){if(pres[3].charAt(j)=='#') slen--;}
            for(int j=0; j<intert/slen; ++j){wholem=wholem+pres[3];}
            int count=intert%slen;
            for(int j=0; count!=0; j++)
            {
                wholem=wholem+pres[3].charAt(j);
                if(pres[3].charAt(j)!='#') count--;
            }
            wholem=wholem+"0";
            int k=-1;
            while(true)
            {
                k=wholem.indexOf(m,k+1);
                if(k==-1) break;
                else if(wholem.charAt(k+m.length())!='#') break; 
            }
            if(k!=-1)
            {
                if(intert>anst || (intert==anst && ansst>stt))
                {
                    answer=pres[2];
                    ansst=stt;
                    anst=intert;
                }
            }
        }
        if(answer=="") answer="(None)";
        return answer;
    }
  }